Teach Elix to keep track of lists of things

stevegrossi opened the pull request:
Resolves #5, but currently a work-in-progress.
### To-Do
- [x] manipulate lists by index
- [x] sanitize list names? (does Redix handle Redis prevent injection?)
- [x] tests (with a mock Redis module, perhaps?)
- [x] Clean up Redis mock (put fixture data in one place?)
- [x] error handling (i.e. adding an item to a nonexistent list, index out of range, etc...)
### Concept

Right now, lists are stored in Redis. I've added responder handlers for creating, deleting, showing, and clearing lists, and adding and removing items from lists.

<img width="349" alt="screen shot 2016-10-16 at 7 14 32 am" src="https://cloud.githubusercontent.com/assets/664341/19417052/54144566-9370-11e6-830e-4f2790c76eae.png">

<img width="398" alt="screen shot 2016-10-16 at 3 11 25 pm" src="https://cloud.githubusercontent.com/assets/664341/19420123/0608389e-93b3-11e6-846a-dfaf9a4d4a8f.png">

Since Redis doesn't allow empty lists (a list is created when its first item is added, and deleted when its first item is removed), and to avoid issues with using complex list names as Redis keys, I'm using two levels of Redis lists for this:
- a "list of lists" to store the names of all lists
- additional lists for each list's items. ~~These will have names like `lists:DIGEST` where `DIGEST` is a hex digest of the name of the list from the "list of lists" above. This avoids invalid Redis keys like `lists:My Favorite Things`.

stevegrossi committed:
Initial attempt at list management

stevegrossi commented on lib/elix/lists.ex:
```diff
@@ -0,0 +1,23 @@
+    all
+  end
+
+end
```
I didn't do this with the GifMeresponder, it felt like a good idea to put the responsibility of actually managing lists into a separate module from the responder, so the responder can just worry about input parsing and delegation.

stevegrossi committed:
List item manipulation

stevegrossi committed:
Minor tweaks

janxious commented:
> validations, e.g. preventing duplicate list names and items

Seems like you could solve this somewhat with using the set commands instead of list commands on redis. (http://redis.io/commands/sadd)

stevegrossi commented:
Indeed, though I think using sets might disallow referencing items by number instead of name, which I think would be handy, since redis sets don't have a defined order or a way to access by index like redis lists do.

stevegrossi committed:
Alright, got some tests going with a mock remix lib

stevegrossi committed:
Resolve credo’s “Pipe chain should start with a raw value”

stevegrossi committed:
One-line moduledoc

stevegrossi committed:
Match all the patterns!

stevegrossi commented:
> - [x] sanitize list names? (does Redix handle Redis prevent injection?)

Sounds like Redis command injection isn't really possible except under some specific circumstances which I don't think apply here: http://stackoverflow.com/questions/22023538/is-it-possible-to-run-a-string-injection-attack-on-a-redis-query

stevegrossi committed:
Operate on lists by number

stevegrossi committed:
Rename Elix.List back to Elix.Lists

Because of the collision with Elixir’s List module, duh

stevegrossi committed:
We probably don’t need hex digests

They just make it harder to tell what’s what

stevegrossi committed:
Docs

stevegrossi committed:
Refactor for DRYness, though I still don’t think…

…I’ve got the right boundaries/responsibilities figured out for these modules

stevegrossi committed:
Prepare for Heroku deploy

janxious commented:
Sorry, trying to get my remotes setup correctly and didn't do it quite right, and ended up deleting the branch. 😸

janxious commented on lib/elix/responders/lists.ex:
```diff
@@ -81,20 +65,31 @@ defmodule Elix.Responders.Lists do
+  end
+
+  defp parse_list_identifier(list_id) do
+    if numeric_string?(list_id) do
```
I think this class might get simpler if `numeric_string?` was a macro and used as a function guard. It might also make it way worse.

stevegrossi replied:
Ooh, I've never written an Elixir macro before. Now would be a good time to learn.

janxious replied:
http://keathley.io/2016/04/09/elixir-guard-clauses.html might be a helpful thing to read first

stevegrossi replied:
Really informative link, but it sounds like using custom macros in guard clauses only works when those macro functions are composed of other functions that would be allowed in guard clauses. Since regexes are not, I don't think a custom macro is possible in this case.

janxious replied:
Having re-read that I agree. Best it would get us is `is_bitstring` which isn't super valuable. I also tried to think of a way to do destructuring with regex, but it wasn't coming together. Here is a wild ride I think would work, but is also a terrible idea:

``` elixir
defp parse_list_identifier(list_id) do
  parse_list_identifier(list_id, "")
end

defp parse_list_identifier([], list_id) do
  list_id
  |> String.to_integer
  |> Lists.get_name
end

defp parse_list_identifier([h|t], list_id) when is_numericstring(h) do
  parse_list_identifier(t, list_id <> h)
end

defp parse_list_identifier(remainder, list_id) do
  list_id <> remainder
end


defmacro is_numericstring(char) do
  quote do
    unquote(char) == "0" ||
      unquote(char) == "1" ||
      unquote(char) == "2" ||
      unquote(char) == "3" ||
      unquote(char) == "4" ||
      unquote(char) == "5" ||
      unquote(char) == "6" ||
      unquote(char) == "7" ||
      unquote(char) == "8" ||
      unquote(char) == "9"
  end
end
```

stevegrossi replied:
That's pretty dang clever! (Which, as you suggest, is reason enough not to use it 😉) After changing the `||`s to `or`s (I also forgot that guard clauses only allow the stricter versions of comparisons) it still doesn't quite work, though I'm sure I could get it working with time.

Instead, I'm looking into possibly eliminating this function entirely and offloading the detection of numeric strings to the regex itself: https://github.com/indyelixir/elix/pull/6/files#r86683827

stevegrossi commented on lib/elix/responders/lists.ex:
```diff
@@ -0,0 +1,105 @@
+  show list <list> - Displays the contents of a list by name or number
+  """
+  respond ~r/show list (.+)/i, %Message{matches: %{1 => list_id}} = msg do
+    list_name = parse_list_identifier(list_id)
```
I'm wondering if I can offload some of the work of `parse_list_identifier` into the regex itself. With named captures, I _think_ I can have the regex key the match as `number` if it's a string of digits, otherwise as `name`. Here's a very naive example that just matches strings of digits vs. strings of non-digits:

``` elixir
iex> Regex.named_captures(~r/test (?<name>\D+)|(?<number>\d+)\Z/, "Gonna test 1234")
%{"name" => "", "number" => "1234"}

iex> Regex.named_captures(~r/test (?<name>\D+)|(?<number>\d+)\Z/, "Gonna test this thing")
%{"name" => "this thing", "number" => ""}
```

Of course, that breaks down with mixed strings:

``` elixir
iex> Regex.named_captures(~r/test (?<name>\D+\Z)|(?<number>\d+\Z)/, "Gonna test this thing 9 times")
nil
```

but the first named capture is naive, matching "strings composed entirely of non-digits" as `name`. What I want to match are "strings not entirely composed of digits". Now, to figure that out at 6am...

stevegrossi replied:
[`(?ifthen|else)` groups](http://www.regular-expressions.info/conditional.html) sound like exactly what I need, but unless I'm making a mistake they don't seem to work in Elixir.

``` elixir
iex> Regex.named_captures(~r/test (?this|99)/, "Gonna test 99")
** (Regex.CompileError) unrecognized character after (? or (?- at position 7
    (elixir) lib/regex.ex:142: Regex.compile!/2
    (elixir) expanding macro: Kernel.sigil_r/2
             iex:6: (file)
```

**Update:** pretty sure I'm making a mistake. From the article, the `ifthen` part needs to be a "condition", not an expression. Time to learn the difference.

stevegrossi replied:
Ok, think I got it. You can turn an expression into a condition with the `=` sign, so with the `(?ifthen|else)` construct, if you make it `(?=ifthen|else)` then it'll match the expression `ifthen`, otherwise attempt to match `else`. Here it is working for me: http://rubular.com/r/RxMGSkLNko

stevegrossi replied:
Then again, `String.to_integer/1`, which we're sometimes using anyway, could also do this for us:

``` elixir
iex> String.to_integer("9999")
9999

iex> String.to_integer("9999 bottles")
** (ArgumentError) argument error
    :erlang.binary_to_integer("9999 bottles")
```

We'd need to `rescue ArgumentError` and return the string in that case, but also an option...

stevegrossi committed:
Eliminate need for `numeric_string?`

stevegrossi committed:
Revert "Eliminate need for `numeric_string?`"

I don't think it's actually better. `rescue ArgumentError` is not at all clear that it applies to `String.to_integer` and might mistakenly rescue other argument errors, since those happen all the time.

This reverts commit dd9087cf1e9dfb7140038cb5259ee02d102ddc05.

stevegrossi committed:
This feels gross, but at least I’m getting valuable test coverage

I don’t like what it’s taking to get the tests passing, but I plan on completely refactoring that later, guided by the tests I’ll have written =/

stevegrossi committed:
:list_not_found and :item_not_found

stevegrossi committed:
Get `with` the program!

Are all these necessary? Probably not, but using `with` everywhere will help me refactor.

stevegrossi committed:
That’s some nicer control flow

Lessons learned: when something has the possibility of not returning an expected value, return a status tuple like `{:ok, value}` from the start.

stevegrossi committed:
Extract `command!` helper

stevegrossi committed:
Fixup doctests

stevegrossi commented on config/dev.exs:
```diff
@@ -4,3 +4,5 @@ config :elix, Elix.Robot,
 
 config :elix, :api_client, Elix.APIClient
+
+config :elix, :redis_client, Redix
```
Since mocks will only ever be used in the test environment, I think it makes sense to declare the actual implementation in `config.exs` so it's inherited by all other environments, rather than having to specify it individually in `dev`, `prod`, etc.

stevegrossi commented on lib/elix.ex:
```diff
@@ -6,11 +6,14 @@ defmodule Elix do
     import Supervisor.Spec, warn: false
 
+    redis_url = System.get_env("REDIS_URL") || "redis://"
+
```
Preparing for Heroku deployment...

stevegrossi commented on lib/elix/responders/lists.ex:
```diff
@@ -0,0 +1,132 @@
+      else
+        {:error, :list_not_found} -> "Sorry, I couldn’t find that list."
+        {:error, :item_not_found} -> "Sorry, I couldn’t find that item."
+      end
```
This really showed me the power of `with`. Before:

``` elixir
response =
  case parse_list_identifier(list_id) do
    :list_not_found ->
        "Sorry, I couldn’t find that list."
    list_name  ->
      case parse_item_identifier(item_id, list_name) do
        :item_not_found ->
          "Sorry, I couldn’t find that item."
        item_name ->
          Lists.delete_item(list_name, item_name)
          render_list(list_name)
      end
  end
```

stevegrossi commented on lib/elix/lists.ex:
```diff
@@ -0,0 +1,136 @@
+    else
+      {:error, :list_not_found}
+    end
+  end
```
The purpose of this function is to check that the list name passed in corresponds to an existing list, but I have some reservations:
- it feels weird having a function that returns its own argument under normal circumstances. I considered writing it as `list_exists?` and returning a boolean, but giving this function the same return values as `get_name` simplifies things in the responder, which now works the same way whether it gets a list index or name.
- I feel like `get_name` and `get_by_name` could be combined somehow. I guess technically because `get_name` is guarded against non-integers, `get_by_name` could become a second body of `get_name` which handles strings. That doesn't feel like a huge improvement, though.
- I'm still not happy with `parse_list_identifier` and `parse_item_identifier` in the responder, which call these functions. Ideally, I'd simplify both somehow.

stevegrossi commented:
Alright, I believe the functionality is now all here. I just pushed some updates to provide helpful error messages when attempting to operate on lists or items which don't exist. Although I still thing there's room for refactoring: I did my best with the error handling code (`with` is awesome!) but I don't think I've made this as simple or clear as it could be. I'll take another look later on.

stevegrossi commented:
I also took [a fun digression](https://github.com/indyelixir/elix/compare/struct) toward treating lists as "things" with an `%Elix.List{}` struct, e.g.

``` elixir

defp render_list(%List{name: name, items: items}) do
  "**#{name}**\n\n#{render_items(items)}"
end
```

This cleaned up the code a bit (pattern-matching on structs is really nice) and felt pleasantly similar to Phoenix's `%Conn{}` structs, etc. but I had two major reservations that led me to abandon that experiment:
- It required unnecessary Redis queries. Since I had the struct include the list's items, that was a second Redis query to build a struct, even for operations that didn't care about the items. The performance impact in this application is negligible and I could possibly have made the items optional (the way Ecto's associations require you to `preload` them), but it still didn't feel right doing unnecessary work.
- It felt weird operating on structs but also having side-effects in Redis, e.g. I'd append a new item to the struct's `items` key, but _also_ insert that item into the appropriate Redis list. I guess that's exactly how Ecto works (you operate on structs but it also persists things to the DB) so it can't be too bad, but I dunno...

As I write this, I'm realizing the struct approach might be worth considering after all. Ultimately, it led to a more Ecto/ActiveRecord-like approach to the list operations: you first get the `%List{}` from the store (which handles errors if not found), and then operate on it.

stevegrossi committed:
Start operating on structs instead of list names

This is still a bit in progress, but I like the direction.

stevegrossi commented:
Just pushed a big commit that starts using `%Elix.List{}` structs and I quite like it. Still some refactoring to be done.

stevegrossi committed:
get_items can be private now

stevegrossi committed:
Fix unused variable

Apparently you can’t match on both sides of the `=`

stevegrossi committed:
Wow, rescue is way faster

https://gist.github.com/stevegrossi/bbca6678f610562d5048415e42f59fe7

stevegrossi committed:
Clean up docs and doctests

stevegrossi committed:
Whoops, RPUSH returns the size of the list

stevegrossi committed:
Move non-test configs into config.exs

We only ever override these in test, so they should be default in other envs.

stevegrossi commented on config/config.exs:
```diff
@@ -21,18 +22,8 @@ config :elix, Elix.Robot,
-#     config :logger, level: :info
-#
+config :elix, :api_client, Elix.APIClient
+config :elix, :redis_client, Redix
```
We only ever override these in `test`, so they should be default in other environmentss.

stevegrossi commented:
I think this turned out pretty good. It could be better, but you gotta move forward...
